common/src/main/java/com/wynntils/core/chat/ChatManager.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.chat;
+
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.mc.event.ChatPacketReceivedEvent;
+import com.wynntils.mc.utils.ComponentUtils;
+import com.wynntils.mc.utils.McUtils;
+import com.wynntils.wc.event.ChatMessageReceivedEvent;
+import com.wynntils.wc.event.NpcDialogEvent;
+import com.wynntils.wc.utils.WynnUtils;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.ChatType;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class ChatManager {
+    private static final Pattern NPC_FINAL_PATTERN =
+            Pattern.compile(" +§[47]Press §r§[cf](SNEAK|SHIFT) §r§[47]to continue§r$");
+
+    private static boolean extractDialog = false;
+    private static String lastRealChat = null;
+    private static List<String> lastNpcDialog = List.of();
+
+    public static void init() {
+        WynntilsMod.getEventBus().register(ChatManager.class);
+    }
+
+    @SubscribeEvent(priority = EventPriority.HIGHEST)
+    public static void onChatReceived(ChatPacketReceivedEvent e) {
+        if (!WynnUtils.onServer()) return;
+        if (e.getType() == ChatType.GAME_INFO) return;
+
+        Component message = e.getMessage();
+        String codedMessage = ComponentUtils.getCoded(message);
+        if (!codedMessage.contains("\n")) {
+            saveLastChat(ComponentUtils.getCoded(message));
+            MessageType messageType = e.getType() == ChatType.SYSTEM ? MessageType.SYSTEM : MessageType.NORMAL;
+            Component updatedMessage = handleChatLine(message, codedMessage, messageType);
+            if (updatedMessage == null) {
+                e.setCanceled(true);
+            } else if (!updatedMessage.equals(message)) {
+                e.setMessage(updatedMessage);
+            }
+            return;
+        }
+
+        if (extractDialog) {
+            handleMultilineMessage(codedMessage);
```

Comment:
This method only handles NPC chat stuff? Is there any more messages with new lines that is not an npc dialogue?

Reply:
This handles the NPC dialog scenario. But when that happens, there are a "pre screen" and "post screen" that does the same thing, but does not contain any of the NPC lines. Kind of a "cls" thing.

I have not seen this system being used for anything else than NPC dialogs. I could not even imagine what that would be. The one thing I know exists but haven't tested it with, is the multi option selections things. I only saw them in Winter Detlas; I don't know if there are any such present in the world today. I imagine the code will work mostly fine anyway, but the dialog itself may be multi-line, which I currently chose not to handle until I see evidence that such a thing exists.

As I said before, I hope Wynncraft prohibits users from posting `\n` to the chat, otherwise all bets are off.